### ISSUE TITLE
Wrong location reported for syntax errors

### ISSUE THREAD

**Issue by asottile** [PyCQA/flake8#740]
In GitLab by @scascketta on May 21, 2020, 14:14

*Please describe how you installed Flake8*

```
$ python3.8 -m venv .venv
$ source .venv/bin/activate
(.venv) $ pip install flake8
```

*Please provide the exact, unmodified output of `flake8 --bug-report`*
```
{
  "dependencies": [],
  "platform": {
    "python_implementation": "CPython",
    "python_version": "3.8.1",
    "system": "Darwin"
  },
  "plugins": [
    {
      "is_local": false,
      "plugin": "mccabe",
      "version": "0.6.1"
    },
    {
      "is_local": false,
      "plugin": "pycodestyle",
      "version": "2.6.0"
    },
    {
      "is_local": false,
      "plugin": "pyflakes",
      "version": "2.2.0"
    }
  ],
  "version": "3.8.1"
}
```
*Please describe the problem or feature*

Running flake8 on a Python module with a syntax error produces the E902 TokenError with the wrong line and column reported.

*If this is a bug report, please explain with examples (and example code) what you expected to happen and what actually happened.*

The contents of the test file with invalid syntax:
```
(.venv) $ cat blah.py
print("hello, world")
)
```

Python reports the correct location of the syntax error:
```
(.venv) $ python3.8 blah.py
  File "blah.py", line 2
    )
    ^
SyntaxError: unmatched ')'
```

Running flake8 on it reports the wrong line and column of the syntax error:
```
(.venv) $ flake8 blah.py
blah.py:1:1: E902 TokenError: EOF in multi-line statement
```

Thank you for reading!

**Comment by asottile** [comment]
In GitLab by @asottile on May 21, 2020, 14:23

tokenization is done before ast parsing so the errors are sometimes different and on different positions:

```console
$ python3 -m tokenize
print("hello, world")
)
1,0-1,5:            NAME           'print'        
1,5-1,6:            OP             '('            
1,6-1,20:           STRING         '"hello, world"'
1,20-1,21:          OP             ')'            
1,21-1,22:          NEWLINE        '\n'           
2,0-2,1:            OP             ')'            
2,1-2,2:            NEWLINE        '\n'           
<stdin>:3:0: error: EOF in multi-line statement
```

though it looks like `TokenError` is always given `1:1` and could maybe be slightly improved (though `3:0` in that example is *also* out of bounds)

**Comment by asottile** [comment]
In GitLab by @reinhrst on Oct 1, 2020, 07:40

I was just looking at this (actually, wrote a question on [stackoverflow](https://stackoverflow.com/questions/64155860/better-alternative-to-flake8s-e902-tokenerror-eof-in-multi-line-statement)).

It makes sense that `tokenize` reports the last line pos 0: that is where it found the `EOF` that it's complaining about.

Also note that if there is a closing bracket missing, `tokenize` just throws the error and does not return anything.

Looking at how pyflake deals with this: they first do an ast parse, since that seems to give a better pointer to where a problem lies:
https://github.com/PyCQA/pyflakes/blob/ef3c5cb31e06e9f094294353c3c83a9decc0fe57/pyflakes/api.py#L40-L58

It's not too hard to hack this into flake8 (however I don't know whether this is something you're willing to have in your code, or whether you prefer it to be somewhere else):

```
   def run_checks(self):
       """Run checks against the file."""
       try:
           self.process_tokens()
           self.run_ast_checks()
       except exceptions.InvalidSyntax as exc:
           import ast
           try:
               ast.parse("".join(self.processor.lines), filename=self.filename)
           except SyntaxError as new_exc:
               self.report(
                   exc.error_code,
                   new_exc.lineno,
                   new_exc.offset - 1,  # seems to be 1-indexed
                   new_exc.msg)
```

With this the example now behaves very nicely indeed:
```
$ flake8 blah.py
test.py:2:2: E902 unmatched ')'
```

And look at how nice the errors become when opening and closing brackets are not matched:
```
$ cat blah.py
print(["hello, world"
       )
$ flake8 blah.py
test.py:2:8: E902 closing parenthesis ')' does not match opening parenthesis '[' on line 1
```

**Comment by asottile** [comment]
In GitLab by @asottile on Oct 1, 2020, 07:54

the fix is simpler than that if my hunch is correct, it should just be swapping the first two lines in that `try:` -- double ast parsing is very very undesirable because of how slow it is

**Comment by asottile** [comment]
In GitLab by @reinhrst on Oct 1, 2020, 08:04

That seems to do the trick as well :). Although in that case it will report both errors (a E999 SyntaxError from the ast part, and the E902 from the tokenize part, the latter still with 1:1 for line/col; so this will still be bad if someone starts fixing their errors from the top of the file); maybe best to add an if() around the `self.process_tokens()`, that it doesn't make sense to do that if the ast doesn't parse (however this may have other side-effects I cannot oversee).

**Comment by asottile** [comment]
In GitLab by @asottile on Oct 1, 2020, 08:07

why would it produce both? wouldn't the SyntaxError exception stop the execution (both don't show up now do they?)

**Comment by asottile** [comment]
In GitLab by @reinhrst on Oct 1, 2020, 08:14

Nope, `run_ast_checks()` calls `self.report()` in the function itself rather than raising it. So the function returns successfully. Might be an idea to have that function raise something as well, and then catch that exception as well in the `except` clause. Doesn't seem like we can reuse `InvalidSyntax` since that's hardcoded to be a `E902`, where ast errors seem to be `E999`.

(On the other hand, might be good to make them E902's)

**Comment by vanschelven** [comment]
I can confirm that the switching of the lines

```
            self.process_tokens()
            self.run_ast_checks()
```
solves the immediate problem of hiding relevant line numbers, by showing both E902 and E999. Admittedly it would be even better if only the E999 would show up in that case

Seems this was introduced in 38ab47702a1cfa451af2e74a95ed145a0f4b2d5a, which manifests itself in the fact that 3.7.7 does not have this problem yet, and 3.7.8 does have it.

**Comment by nyuszika7h** [comment]
I still see this issue with flake8 3.9.2. I end up having to actually run the code in the Python interpreter to get a SyntaxError with the correct line number.

Simple example:
```py
foo = 42    
bar = (1 + 1
baz = 'qux' 
```

```
foo.py|1 col 1 error| E902: TokenError: EOF in multi-line statement
```

**Comment by asottile** [comment]
try master

**Comment by nyuszika7h** [comment]
Oh yeah, master works. I assumed it was already included in the 3.9.2 release based on the commit history.

**Comment by asottile** [comment]
this doesn't mention any tag: https://github.com/PyCQA/flake8/commit/d1a4043c592c7a61bb65d76dfe98be4a356414ae

and if you run:

```console
$ git describe --contains d1a4043c592c
fatal: cannot describe 'd1a4043c592c7a61bb65d76dfe98be4a356414ae'
```

it has not been released (it will be in 4.x)
